tests/integration/helpers.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# Apache Software License, version 2.0
+
+"""Helpers for the slurmctld integration tests."""
+
+import contextlib
+import logging
+import pathlib
+import shlex
+import subprocess
+import tempfile
+from io import StringIO
+from typing import Dict
+
+import paramiko
+import tenacity
+from pytest_operator.plugin import OpsTest
+from urllib import request
+
+logger = logging.getLogger(__name__)
+
+ETCD = "etcd-v3.5.0-linux-amd64.tar.gz"
+ETCD_URL = f"https://github.com/etcd-io/etcd/releases/download/v3.5.0/{ETCD}"
+NHC = "lbnl-nhc-1.4.3.tar.gz"
+NHC_URL = f"https://github.com/mej/nhc/releases/download/1.4.3/{NHC}"
+SINGULARITY_DEB = "singularity-ce_3.10.2-focal_amd64.deb"
+SINGULARITY_DEB_URL = (
+    f"https://github.com/sylabs/singularity/releases/download/v3.10.2/{SINGULARITY_DEB}"
+)
+SINGULARITY_RPM = "singularity-ce-3.10.2-1.el7.x86_64.rpm"
+SINGULARITY_RPM_URL = (
+    f"https://github.com/sylabs/singularity/releases/download/v3.10.2/{SINGULARITY_RPM}"
+)
+VERSION = "version"
+VERSION_NUM = subprocess.run(
+    shlex.split("git describe --always"), stdout=subprocess.PIPE, text=True
+).stdout.strip("\n")
+
+
+def get_slurmctld_res() -> Dict[str, pathlib.Path]:
+    """Get slurmctld resources needed for charm deployment."""
+    if not (version := pathlib.Path(VERSION)).exists():
+        logger.info(f"Setting resource {VERSION} to value {VERSION_NUM}...")
+        version.write_text(VERSION_NUM)
+    if not (etcd := pathlib.Path(ETCD)).exists():
+        logger.info(f"Getting resource {ETCD} from {ETCD_URL}...")
+        request.urlretrieve(ETCD_URL, etcd)
+
+    return {"etcd": etcd}
+
+
+def get_slurmd_res() -> Dict[str, pathlib.Path]:
+    """Get slurmd resources needed for charm deployment."""
+    if not (nhc := pathlib.Path(NHC)).exists():
+        logger.info(f"Getting resource {NHC} from {NHC_URL}...")
+        request.urlretrieve(NHC_URL, nhc)
+    if not (singularity_deb := pathlib.Path(SINGULARITY_DEB)).exists():
+        logger.info(f"Getting resource {SINGULARITY_DEB} from {SINGULARITY_DEB_URL}...")
+        request.urlretrieve(SINGULARITY_DEB_URL, singularity_deb)
+    if not (singularity_rpm := pathlib.Path(SINGULARITY_RPM)).exists():
+        logger.info(f"Getting resource {SINGULARITY_RPM} from {SINGULARITY_RPM_URL}...")
+        request.urlretrieve(SINGULARITY_RPM_URL, singularity_rpm)
+
+    return {"nhc": nhc, "singularity-deb": singularity_deb, "singularity-rpm": singularity_rpm}
+
+
+@contextlib.asynccontextmanager
```

Comment:
I assume that you are talking about the asynchronous context manager? If so, I find that it's just another way to peel the same potato. My other reasoning for using this context manager is that if I am testing locally (my workstation rather than GitHub Actions), I am using temporary ssh keys rather than my main ssh key on my system. Also, I just like it because you can control the implementation + you can chain/pipe commands and use wildcards inside the unit rather on the host system.

I think the asynchronous context manager will be useful to have in the future (especially for complicated tests), but I would like to hear your thoughts further.

Reply:
Another thing. When testing locally, my ssh key set up with Juju requires a password. This will bork the automatic flow of the tests.

Reply:
Resolved in https://github.com/canonical/slurmctld-operator/pull/1/commits/975e4a3d1548d6df577d08572406eb25c07639d9. See comment here for link to the public asynchronous context manager gist: https://github.com/canonical/slurmdbd-operator/pull/4#discussion_r1085617719.